File: frontend/src/utils/privacy.js
export function canViewFullBidderUsername(role) {
  return ['admin', 'defender'].includes(String(role || '').toLowerCase())
}

export function maskUsername(username) {
  if (!username) {
    return username
  }

  const value = String(username)
  if (value.includes('*')) {
    return value
  }

  if (value.length <= 2) {
    return `${value[0] || ''}${'*'.repeat(Math.max(value.length - 1, 0))}`
  }

  if (value.length <= 4) {
    return `${value[0]}${'*'.repeat(value.length - 2)}${value.at(-1)}`
  }

  const suffixLength = value.length >= 8 ? 2 : 1
  return `${value.slice(0, 2)}${'*'.repeat(Math.max(value.length - 2 - suffixLength, 1))}${value.slice(-suffixLength)}`
}

export function formatBidderLabel(username, bidderId, role) {
  if (username) {
    return `@${canViewFullBidderUsername(role) ? username : maskUsername(username)}`
  }

  return bidderId ? `Bidder ${bidderId}` : 'Bidder'
}

File: backend/app/utils/privacy.py
PRIVILEGED_BID_VIEW_ROLES = {"admin", "defender"}


def normalize_role(role) -> str:
    value = getattr(role, "value", role)
    return str(value or "").strip().lower()


def can_view_full_bidder_username(user) -> bool:
    return normalize_role(getattr(user, "role", None)) in PRIVILEGED_BID_VIEW_ROLES


def mask_username(username: str | None) -> str | None:
    if not username:
        return username

    value = str(username)
    length = len(value)

    if length <= 2:
        return value[0] + "*" * max(length - 1, 0)

    if length <= 4:
        return f"{value[0]}{'*' * (length - 2)}{value[-1]}"

    prefix_length = 2
    suffix_length = 2 if length >= 8 else 1
    masked_length = max(length - prefix_length - suffix_length, 1)
    return f"{value[:prefix_length]}{'*' * masked_length}{value[-suffix_length:]}"


def public_bidder_username(username: str | None, viewer=None) -> str | None:
    if can_view_full_bidder_username(viewer):
        return username
    return mask_username(username)
